libexec/honest-clone

```diff
@@ -1,83 +1,100 @@
+source "$SCRIPT_DIR"/../libexec/honest-common.sh
 
-usage() {
-  sed -ne '/^#/!q;s/.\{1,2\}//;1,2d;p' < "$0"
-  [ -z "$1" ] || exit "$1"
-}
+#######################################
+# Convert <vendor>:<author>/<project>[@..] to https://<vendor's url>/<author>/<project>[@..]
+# When failed to convert the format, the original string will be return.
+# Globals:
+#   None
+# Arguments:
+#   <Repo> The repo name with format '<vendor>:<author>/<project>[@..]'
+# Returns:
+#   None
+#   Echos the transformed string
+#######################################
+abbr_form() {
+  local vendor=${1%%:*}
+  local repo=${1#*:}
+  local index=$(get_index "$vendor" "${REPO_VENDORS[@]}")
 
-is_start_with() {
-  [[ "$1" != "$2"* ]] && return 1
-  local str="$1"
-  RESULT="${str:${#2}}"
+  if [[ "$1" == "https://"* ]]; then
+    # Return the original string if it's already in the format
+    echo "$1"
+  elif [[ "$index" != "-1" ]]; then
+    # Return the URL if found
+    echo "https://${REPO_VENDORS_HOST[$index]}/$repo"
+  else
+    # Return the original string if fail to parse
+    echo "$1"
+  fi
+  return 0
 }
 
-abbr_form() {
-  # check if url in form <vendor>:<author>/<project>[@..]
-  # if yes, convert it to https://<vendor's url>/<author>/<project>[@..]
-  local url="$1"
-  local vendor_map=("github" "github.com" "bitbucket" "bitbucket.org" "gitlab" "gitlab.com")
-  local n=${#vendor_map[@]}
-  for (( i=0; i<=$n; i+=2 )); do
-    is_start_with "$url" "${vendor_map[$i]}:"
-    if [[ $? -eq 0 ]]; then
-      RESULT="https://${vendor_map[$i+1]}/$RESULT"
-      return 0
-    fi
-  done
-  return 1
+#######################################
+# Parse the tag string from git URL string which can be
+# 'https://<author>@github.com/<authors>/<repo>@v1.00'.
+# The tag string will be 'v1.00' in this case.
+# Globals:
+#   None
+# Arguments:
+#   <Repo> The repo name with format '<vendor>:<author>/<project>[@..]'
+# Returns:
+#   None
+#   Echos the tag string; empty if not found
+#######################################
+get_tag_from_url() {
+  local url=${1#*//} # Remove string before //
+  local proj_path=${url#*/} # Remove string before the first /
+  local tag=$([[ "$proj_path" == *"@"* ]] && echo ${proj_path##*@} || echo "")
+
+  echo $tag
 }
 
-get_tag() {
-  # scan for first '@'
-  # therefore, we don't support urls like 'https://author@github.com/author/proj'
+git_clone() {
   local url="$1"
-  local n=${#url}
-  for (( i=0; i<=$n; i++ )); do
-    if [[ ${url:$i:1} == '@' ]]; then
-      REPO=${url:0:$i}
-      TAG=${url:$i+1}
-    fi
-  done
-  return 0
+  local tag="$2"
+  local path="$3"
+  set -e
+  # TODO Use shallow clone or other techniques to improve the speed
+  git clone "$url" "$path" >/dev/null
+  cd $path
+  # if $tag is not present, checkout to the latest release.
+  tag="${tag:-$(git describe --tags `git rev-list --tags --max-count=1 2>/dev/null` 2>/dev/null)}"
+  [[ "$tag" != "" ]] && git checkout --detach "$tag" >/dev/null
```

Comment:
I think it's ok to show the checking out message to users (no need >/dev/null)
Or maybe we can add silent/verbose options to control the logging level